File: backend/internal/models/rooms.go
package models

type Room struct {
	ID           string `json:"id"`
	RoomNumber   int    `json:"room_number"`
	Floor        int    `json:"floor"`
	SuiteType    string `json:"suite_type"`
	RoomStatus   string `json:"room_status"`
	IsAccessible bool   `json:"is_accessible"`
} //@name Room

type RoomSortOption string

const (
	RoomSortAscending  RoomSortOption = "ascending"
	RoomSortDescending RoomSortOption = "descending"
	RoomSortUrgency    RoomSortOption = "urgency"
)

type FilterRoomsRequest struct {
	Floors     *[]int         `json:"floors,omitempty"     validate:"omitempty,dive,min=1"`
	Limit      int            `json:"limit,omitempty"      validate:"min=0"`
	Cursor     string         `json:"cursor,omitempty"     validate:"omitempty"`
	Status     []string       `json:"status,omitempty"`     // occupied | vacant | open-tasks
	Attributes []string       `json:"attributes,omitempty"` // standard | deluxe | suite | accessible
	Advanced   []string       `json:"advanced,omitempty"`   // arrivals-today | departures-today
	Sort       RoomSortOption `json:"sort,omitempty"`
} //@name FilterRoomsRequest

// Read model for rooms page on the frontend
type RoomWithOptionalGuestBooking struct {
	Room
	Guests             []Guest         `json:"guests"`
	BookingStatus      BookingStatus   `json:"booking_status"`
	Priority           RequestPriority `json:"priority"`
	HasUnassignedTasks bool            `json:"has_unassigned_tasks"`
} //@name RoomWithOptionalGuestBooking

File: clients/shared/src/api/rooms.ts
import { useQuery } from "@tanstack/react-query";
import { GithubComGenerateSelfserveInternalModelsBookingStatus } from "./generated/models";
import type { RoomWithOptionalGuestBooking } from "./generated/models";
import { useAPIClient } from "./client";

export const BookingStatus = GithubComGenerateSelfserveInternalModelsBookingStatus;

export const RoomStatusValue = {
  Available: "available",
  Cleaning: "cleaning",
  OutOfOrder: "out-of-order",
} as const;

export type RoomStatusFilter = "occupied" | "vacant" | "open-tasks";
export type RoomAttributeFilter = "standard" | "deluxe" | "suite" | "accessible";
export type RoomAdvancedFilter = "arrivals-today" | "departures-today";
export type RoomSortOption = "ascending" | "descending" | "urgency";

export type RoomFiltersParams = {
  floors?: number[];
  status?: RoomStatusFilter[];
  attributes?: RoomAttributeFilter[];
  advanced?: RoomAdvancedFilter[];
  sort?: RoomSortOption;
  cursor?: string;
  limit?: number;
};

type RoomsPage = {
  items: RoomWithOptionalGuestBooking[] | null;
  has_more: boolean;
  next_cursor: string | null;
};

export const getRoomsQueryKey = (params: RoomFiltersParams) =>
  ["rooms", params] as const;

export const useGetRooms = (params: RoomFiltersParams, enabled = true) => {
  const api = useAPIClient();
  return useQuery({
    queryKey: getRoomsQueryKey(params),
    queryFn: () =>
      api.post<RoomsPage>("/rooms", { limit: 200, ...params }),
    enabled,
  });
};

export const useGetRoomsFloors = () => {
  const api = useAPIClient();
  return useQuery({
    queryKey: ["rooms", "floors"],
    queryFn: () => api.get<number[]>("/rooms/floors"),
  });
};
